### DISCUSSION TITLE
Colormaps and colorbars with single color and varying alpha

### DISCUSSION THREAD

**Discussion by ianthomas23** [bokeh/bokeh#12202] [Internals and design (Q&A)] [answered]
I would like to support colormaps/colorbars which use a single color but varying alpha. There are 2 use cases:

  1. Datashader uses this approach for categorical data with a single color per category and an alpha that depends on the amount of that category or the total of all categories.
  2. Contouring of cloud cover and pollution densities on top of maps, so the map is partially visible through the cloud/pollution.

As e.g. `fill_color` and `fill_alpha` are separate properties it is not sensible to add this functionality to the existing `ColorMapper` classes. So I propose to add new `ScalarMapper` classes which do pretty much the same as the `ColorMapper` classes but they map from data space to a scalar value rather than to a color. They could be used for other things than `alpha`, hence the more generic name. We would need separate `LinearScalarMapper`, `LogScalarMapper` and `EqHistScalarMapper` and from the user's point of view they could be hooked up to any scalar visual property just like the `ColorMapper`s are hooked up to colors.

`ColorMapper` takes a sequence of colors. `ScalarMapper` could take a sequence of scalars but I think it more sensible for it to take `start` and `end` values (names TBD) to interpolate between. Probably it would have to accept an integer number of steps to be analogous to the number of palette entries for `ColorMapper`, but this could be optional if you wanted full-precision interpolation rather than the binned interpolation of `ColorMapper`.

Adding this to `ColorBar` isn't quite as clear. It currently takes a `ColorMapper` so we'd have to add an optional `ScalarMapper` for the `alpha` and either support a `ColorMapper` of a single value (this might work already?) or allow a single color to be specified as an alternative to the `ColorMapper`. Given that `ColorBar` returns a sequence of RGBA colors, the display of the colorbar shouldn't be a problem although we'd need to cope with different numbers of entries in the `ColorMapper` and `ScalarMapper`.

**Comment by mattpap** [comment, marked as answer]
> As e.g. fill_color and fill_alpha are separate properties (...)

Just to note, `fill_color` can contain alpha channel now (since we allowed support for CSS4 colors).

**Reply by ianthomas23** [reply]
Thanks, I didn't know that. Which means I can accomplish everything I want using the existing colormapper and colorbar classes.
